replica/table.cc

```diff
@@ -2088,9 +2104,41 @@ future<> tablet_storage_group_manager::update_effective_replication_map(const lo
     if (new_tablet_count > old_tablet_count) {
         tlogger.info0("Detected tablet split for table {}.{}, increasing from {} to {} tablets",
                         schema()->ks_name(), schema()->cf_name(), old_tablet_count, new_tablet_count);
-        return handle_tablet_split_completion(*old_tablet_map, *new_tablet_map);
+        co_await handle_tablet_split_completion(*old_tablet_map, *new_tablet_map);
+        co_return;
     }
-    return make_ready_future();
+
+    // Allocate/deallocate storage group if tablet is migrating in/out.
+    auto this_replica = locator::tablet_replica{
+        .host = erm.get_token_metadata().get_my_id(),
+        .shard = this_shard_id()
+    };
+    auto tablet_migrates_in = [this_replica] (locator::tablet_transition_info& transition_info) {
+        return transition_info.stage == locator::tablet_transition_stage::allow_write_both_read_old && transition_info.pending_replica == this_replica;
+    };
+    auto tablet_migrates_out = [&new_tablet_map, this_replica] (locator::tablet_id tid, locator::tablet_transition_info& transition_info) {
+        auto& current = new_tablet_map->get_tablet_info(tid).replicas;
+        auto is_this_replica = [this_replica] (const locator::tablet_replica& replica) { return replica == this_replica; };
+        return transition_info.stage == locator::tablet_transition_stage::end_migration &&
```

Comment:
Tablet may migrate out also in `revert_migration` stage.

But I wonder if we have to make so much assumptions here about the state machine workings (better not to). We could scan all tablet tids and check if tablet_map::has_replica_on(tid, this_replica), and if not and cg allocated => drop, if yes and cg not allocated => allocate.